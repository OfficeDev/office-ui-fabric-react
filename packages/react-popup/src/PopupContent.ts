export * from './components/PopupContent/index';
