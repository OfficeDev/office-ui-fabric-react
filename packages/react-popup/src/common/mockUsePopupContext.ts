import { PopupContextValue, usePopupContext } from '../popupContext';

/**
 * A test utility to mock the usePopupContext hook that relies on context selector
 * Don't forget to call `jest.mock(**\/popupContext.ts)` while using this
 * @param options Popup context values to set for testing
 */
export const mockPopupContext = (options: Partial<PopupContextValue> = {}) => {
  const mockContext: PopupContextValue = {
    target: null,
    triggerRef: { current: null },
    contentRef: { current: null },
    open: false,
    setOpen: jest.fn(),
    mountNode: null,
    openOnContext: false,
    openOnHover: false,
    ...options,
  };
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  (usePopupContext as jest.Mock).mockImplementation((selector: (context: PopupContextValue) => any) => {
    return selector(mockContext);
  });
};
