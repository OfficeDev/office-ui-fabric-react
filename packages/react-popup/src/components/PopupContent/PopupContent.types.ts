import * as React from 'react';
import { ComponentProps, ComponentState } from '@fluentui/react-utilities';
import { PopupContextValue } from '../../popupContext';

/**
 * PopupContent Props
 */
export interface PopupContentProps extends ComponentProps, React.HTMLAttributes<HTMLElement> {}

/**
 * Names of the shorthand properties in PopupContentProps
 */
export const popupContentShorthandProps = [] as const;

/**
 * Names of the shorthand properties in PopupContentProps
 */
export type PopupContentShorthandProps = typeof popupContentShorthandProps[number];

/**
 * Names of PopupContentProps that have a default value in usePopupContent
 */
export type PopupContentDefaultedProps = never;

/**
 * PopupContent State
 */
export type PopupContentState = ComponentState<
  React.Ref<HTMLElement>,
  PopupContentProps & Pick<PopupContextValue, 'open' | 'mountNode'>,
  PopupContentShorthandProps,
  PopupContentDefaultedProps
>;
