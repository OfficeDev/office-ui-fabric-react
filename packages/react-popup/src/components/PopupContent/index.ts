export * from './PopupContent';
export * from './PopupContent.types';
export * from './renderPopupContent';
export * from './usePopupContent';
export * from './usePopupContentStyles';
