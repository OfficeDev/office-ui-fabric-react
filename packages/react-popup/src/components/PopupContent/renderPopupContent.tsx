import * as React from 'react';
import { getSlots } from '@fluentui/react-utilities';
import { Portal } from '@fluentui/react-portal';
import { popupContentShorthandProps, PopupContentState } from './PopupContent.types';

/**
 * Render the final JSX of PopupContent
 */
export const renderPopupContent = (state: PopupContentState) => {
  const { slots, slotProps } = getSlots(state, popupContentShorthandProps);

  // TODO should hidden popups be supported ?
  if (!state.open) {
    return null;
  }

  return (
    <Portal mountNode={state.mountNode}>
      <slots.root {...slotProps.root}>{state.children}</slots.root>
    </Portal>
  );
};
