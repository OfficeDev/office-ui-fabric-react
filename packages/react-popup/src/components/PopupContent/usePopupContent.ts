import * as React from 'react';
import { makeMergeProps, resolveShorthandProps, useMergedRefs } from '@fluentui/react-utilities';
import { PopupContentProps, popupContentShorthandProps, PopupContentState } from './PopupContent.types';
import { usePopupContext } from '../../popupContext';

const mergeProps = makeMergeProps<PopupContentState>({ deepMerge: popupContentShorthandProps });

/**
 * Create the state required to render PopupContent.
 *
 * The returned state can be modified with hooks such as usePopupContentStyles,
 * before being passed to renderPopupContent.
 *
 * @param props - props from this instance of PopupContent
 * @param ref - reference to root HTMLElement of PopupContent
 * @param defaultProps - (optional) default prop values provided by the implementing type
 */
export const usePopupContent = (
  props: PopupContentProps,
  ref: React.Ref<HTMLElement>,
  defaultProps?: PopupContentProps,
): PopupContentState => {
  const contentRef = usePopupContext(context => context.contentRef);
  const open = usePopupContext(context => context.open);
  const openOnContext = usePopupContext(context => context.openOnContext);
  const openOnHover = usePopupContext(context => context.openOnHover);
  const setOpen = usePopupContext(context => context.setOpen);
  const mountNode = usePopupContext(context => context.mountNode);

  const state = mergeProps(
    {
      open: false,
      mountNode,
      ref: useMergedRefs(ref, contentRef),
    },
    defaultProps && resolveShorthandProps(defaultProps, popupContentShorthandProps),
    resolveShorthandProps(props, popupContentShorthandProps),
  );

  state.open = open || false;

  const { onMouseEnter: onMouseEnterOriginal, onMouseLeave: onMouseLeaveOriginal } = state;
  state.onMouseEnter = (e: React.MouseEvent<HTMLElement>) => {
    if (openOnHover && !openOnContext) {
      setOpen(e, true);
    }

    onMouseEnterOriginal?.(e);
  };

  state.onMouseLeave = (e: React.MouseEvent<HTMLElement>) => {
    if (openOnHover && !openOnContext) {
      setOpen(e, false);
    }

    onMouseLeaveOriginal?.(e);
  };

  return state;
};
