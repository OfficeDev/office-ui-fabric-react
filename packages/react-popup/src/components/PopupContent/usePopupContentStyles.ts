import { makeStyles, mergeClasses } from '@fluentui/react-make-styles';
import { PopupContentState } from './PopupContent.types';

/**
 * Styles for the root slot
 */
const useStyles = makeStyles({
  root: theme => ({
    backgroundColor: theme.alias.color.neutral.neutralBackground1,
    boxShadow: theme.alias.shadow.shadow16,
  }),
});

/**
 * Apply styling to the PopupContent slots based on the state
 */
export const usePopupContentStyles = (state: PopupContentState): PopupContentState => {
  const styles = useStyles();
  state.className = mergeClasses(styles.root, state.className);

  return state;
};
