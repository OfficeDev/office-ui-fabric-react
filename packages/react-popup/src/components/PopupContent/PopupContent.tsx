import * as React from 'react';
import { usePopupContent } from './usePopupContent';
import { PopupContentProps } from './PopupContent.types';
import { renderPopupContent } from './renderPopupContent';
import { usePopupContentStyles } from './usePopupContentStyles';

/**
 * PopupContent component renders react children in a positioned box
 */
export const PopupContent = React.forwardRef<HTMLElement, PopupContentProps>((props, ref) => {
  const state = usePopupContent(props, ref);

  usePopupContentStyles(state);
  return renderPopupContent(state);
});

PopupContent.displayName = 'PopupContent';
