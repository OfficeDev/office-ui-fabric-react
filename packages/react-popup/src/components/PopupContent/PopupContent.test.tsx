import * as React from 'react';
import { PopupContent } from './PopupContent';
import { render, fireEvent } from '@testing-library/react';
import { ReactWrapper } from 'enzyme';
import { isConformant } from '../../common/isConformant';
import { Portal } from '@fluentui/react-portal';
import { mockPopupContext } from '../../common/mockUsePopupContext';

jest.mock('../../popupContext');

describe('PopupContent', () => {
  isConformant({
    Component: PopupContent,
    displayName: 'PopupContent',
    requiredProps: { open: true },
    helperComponents: [Portal],
  });

  let wrapper: ReactWrapper | undefined;

  afterEach(() => {
    if (wrapper) {
      wrapper.unmount();
      wrapper = undefined;
    }
  });

  beforeEach(() => {
    mockPopupContext({ open: true });
  });

  // PopupContent is rendered by a Portal so won't be available in the rendered container
  const testid = 'component';

  /**
   * Note: see more visual regression tests for PopupContent in /apps/vr-tests.
   */
  it('renders a default state', () => {
    const { getByTestId } = render(<PopupContent data-testid={testid}>Default PopupContent</PopupContent>);
    expect(getByTestId(testid)).toMatchSnapshot();
  });

  it.each([
    ['onMouseEnter', fireEvent.mouseEnter],
    ['onMouseLeave', fireEvent.mouseLeave],
  ])('should keep the original %s handler', (handler, triggerEvent) => {
    // Arrange
    const spy = jest.fn();
    const { getByTestId } = render(
      <PopupContent data-testid={testid} {...{ [handler]: spy }}>
        Content
      </PopupContent>,
    );

    // Act
    triggerEvent(getByTestId(testid));

    // Assert
    expect(spy).toHaveBeenCalledTimes(1);
  });

  it('should not render when open is false', () => {
    // Arrange
    mockPopupContext({ open: false });
    const { queryByTestId } = render(<PopupContent data-testid={testid}>Content</PopupContent>);

    // Assert
    expect(queryByTestId(testid)).toBeNull();
  });
});
