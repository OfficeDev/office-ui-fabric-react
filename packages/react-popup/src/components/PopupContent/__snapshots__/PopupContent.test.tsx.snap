// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PopupContent renders a default state 1`] = `
<div
  class=""
  data-testid="component"
  role="dialog"
>
  Default PopupContent
</div>
`;
